tensorflow/core/ops/candidate_sampling_ops.cc

```diff
@@ -373,22 +374,23 @@ seed2: An second seed to avoid seed collision.
 REGISTER_OP("ComputeAccidentalHits")
     .Input("true_classes: int64")
     .Input("sampled_candidates: int64")
+    .Input("num_true: int32")
     .Output("indices: int32")
     .Output("ids: int64")
     .Output("weights: float")
-    .Attr("num_true: int")
     .Attr("seed: int = 0")
     .Attr("seed2: int = 0")
     .SetShapeFn([](InferenceContext* c) {
-      int64 num_true;
```

Comment:
So the shape inference should still work when we have a constant? Is that code done?

Reply:
This is an upper mentioned check that I left out in `ComputeAccidentalHits`. Theoretically the commented code should work, but I would get a `Segmentation fault (core dumped)` when `ComputeAccidentalHits` is invoked.

Reply:
This approach worked well for a sampler [code](https://github.com/MtDersvan/tensorflow/blob/r1.4/tensorflow/core/kernels/candidate_sampler_ops.cc#L48), and `CandidateSamplerShapeFn` works with `num_true` as an Input, as well.
I tried two methods:

1) 
~~~c++
int32 num_true = c->input_tensor(2)->scalar<int32>()();
~~~

With a check:
~~~c++
DimensionHandle unused;
TF_RETURN_IF_ERROR(c->WithValue(c->Dim(true_classes, 1), num_true, &unused));
~~~

2) 
~~~c++
DimensionHandle num_true;
TF_RETURN_IF_ERROR(c->MakeDimForScalarInput(2, &num_true));
~~~

With a check:
~~~c++
DimensionHandle unused;
TF_RETURN_IF_ERROR(c->WithValue(c->Dim(true_classes, 1), c->Value(num_true), &unused));
~~~

But both yielded `Segmentation fault (core dumped)` error, when python `compute_accidental_hits` op is run.
Maybe I'm doing something wrong? I just wanted to get a scalar representation of a `num_true` tensor, to use it in a `WithValue` func.

Reply:
Are you running with a GPU build? Maybe set HostMemory?

Reply:
-1 is the shape when it is unknown. You have to support that. It's usually because it is using a different path which loses the shape information. If you could figure out which one it is, that would be great.

Reply:
@drpngx sorry for being absent for so long, but I'm back. What's the best way to track the path?
This original part:
~~~c++
        int64 num_true;
        TF_RETURN_IF_ERROR(c->GetAttr("num_true", &num_true));
        ...
        DimensionHandle unused;
        TF_RETURN_IF_ERROR(
            c->WithValue(c->Dim(true_classes, 1), num_true, &unused));
~~~
Was substituted by this one:
~~~c++
        DimensionHandle num_true;
        TF_RETURN_IF_ERROR(c->MakeDimForScalarInput(2, &num_true));
        ...
        DimensionHandle unused;
        TF_RETURN_IF_ERROR(
            c->WithValue(c->Dim(true_classes, 1), c->Value(num_true), &unused));
~~~
And that change gives this error: https://pastebin.com/r5iTHZ6L

Also, the upper mentioned code, looks more like a hack, but when I try to get the value directly:
~~~c++
      int64 num_true;
      const Tensor* t = c->input_tensor(2);
      TF_RETURN_IF_ERROR(c->GetScalarFromTensor(t, &num_true));
~~~
It gives me a `Segmentation fault (core dumped)` when I launch `nce_loss` with `remove_accidental_hits=True`.

I'll try a few more things, but let me know if I'm making some terrible mistakes.

Reply:
Surprisingly this:
~~~c++
REGISTER_OP("ComputeAccidentalHits")
    .Input("true_classes: int64")
    .Input("sampled_candidates: int64")
    .Input("num_true: int32")
    .Output("indices: int32")
    .Output("ids: int64")
    .Output("weights: float")
    .Attr("seed: int = 0")
    .Attr("seed2: int = 0")
    .SetShapeFn([](InferenceContext* c) {
      DimensionHandle num_true;
      TF_RETURN_IF_ERROR(c->MakeDimForScalarInput(2, &num_true));

      // Validate true_classes.
      ShapeHandle true_classes;
      TF_RETURN_IF_ERROR(c->WithRank(c->input(0), 2, &true_classes));
      DimensionHandle unused;
      TF_RETURN_IF_ERROR(
          c->WithValue(c->Dim(true_classes, 1), c->Value(num_true), &unused));

      // All three outputs are the same shape.
      ShapeHandle v = c->Vector(InferenceContext::kUnknownDim);
      c->set_output(0, v);
      c->set_output(1, v);
      c->set_output(2, v);
      return Status::OK();
    })
~~~
works fine when `Value(Dim(true_classes, 1)) == kUnknownDim`, e.g. when both `num_true` is a placeholder and `labels'` last dimension is a placeholder.
Snippet: https://pastebin.com/fCgNwwYY

Reply:
btw, I tried to add `HostMemory`:
~~~c++
REGISTER_KERNEL_BUILDER(
    Name("ComputeAccidentalHits")
    .Device(DEVICE_CPU)
    .HostMemory("num_true"),
    ComputeAccidentalHitsOp);
~~~
But I am still getting `Segmentation fault (core dumped)`, any time I try to access tensor data via `c->input_tensor(2);` 
I think this is the main blocker.

Also, I think there needs to be an additional handling case a side from `WithValue` when `num_true` is undefined or a placeholder.